oak/server/rust/oak_runtime/src/node/http/mod.rs

```diff
@@ -0,0 +1,520 @@
+            invocation_channel
+        );
+        Ok(invocation_channel)
+    }
+
+    fn get_invocation_channel(
+        runtime: &RuntimeProxy,
+        startup_handle: oak_abi::Handle,
+    ) -> Result<oak_abi::Handle, OakStatus> {
+        runtime
+                .channel_read(startup_handle)
+                .map_err(|error| {
+                    error!("Couldn't read from the initial reader handle {:?}", error);
+                    OakStatus::ErrInternal
+                })
+                .and_then(|message| {
+                    message
+                        .ok_or_else(|| {
+                            error!("Empty message");
+                            OakStatus::ErrInternal
+                        })
+                        .and_then(|m| {
+                            if m.handles.len() == 1 {
+                                let handle = m.handles[0];
+                                match runtime.channel_direction(handle)? {
+                                    ChannelHalfDirection::Write => Ok(handle),
+                                    ChannelHalfDirection::Read => {
+                                        error!(
+                                            "Http server pseudo-node should receive a writer handle, found reader handle {}",
+                                            handle
+                                        );
+                                        Err(OakStatus::ErrBadHandle)
+                                    },
+                                }
+                            } else {
+                                error!(
+                                    "Http server pseudo-node should receive a single writer handle, found {}",
+                                    m.handles.len()
+                                );
+                                Err(OakStatus::ErrInternal)
+                            }
+                        })
+                })
+    }
+
+    // Make a server, with graceful shutdown, from the given [`HttpRequestHandler`].
+    async fn make_server(
+        &self,
+        request_handler: HttpRequestHandler,
+        notify_receiver: tokio::sync::oneshot::Receiver<()>,
+    ) {
+        // A `Service` is needed for every connection, so this
+        // creates one from the `request_handler`.
+        let make_service = make_service_fn(move |_conn| {
+            let handler = request_handler.clone();
+            async move {
+                Ok::<_, hyper::Error>(service_fn(move |req| {
+                    let handler = handler.clone();
+
+                    async move {
+                        let http_request = HttpServerNode::map_to_http_request(req).await;
+                        handler.handle(http_request).await
+                    }
+                }))
+            }
+        });
+
+        let server = Server::bind(&self.address).serve(make_service);
+        let graceful = server.with_graceful_shutdown(async {
```

Comment:
This is a server, not a graceful :) (though I realise that the documentation uses that naming: https://docs.rs/hyper/0.13.7/hyper/server/struct.Server.html#method.with_graceful_shutdown). Maybe this can be inlined with the builder on the previous line? Or just shadow `server` itself? not sure